### PR TITLE
Fix Documentation Javadoc Build

### PR DESCRIPTION
**Pull request by awholegunch**
Fix errors in pom-xml files. Revises documentation Javadoc build to use a local repo, and install
CDAP there. Split Javadocs from first part of command to aid debugging.

Modifies doc build script so that errors in Javadoc build are fatal and are surfaced immediately.

Passes a full doc build: https://builds.cask.co/browse/CDAP-DBT12-1

New Javadocs: https://builds.cask.co/artifact/CDAP-DBT12/shared/build-1/Current-Docs-HTML/3.5.5/en/reference-manual/javadocs/index.html

Once I fixed both of these warnings (see below) I was able to get the local repo to work, which kept failing until then.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml [hunk 1]
```diff
@@ -71,7 +71,9 @@
     </resources>
     <plugins>
       <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-shade-plugin</artifactId>
+        <version>2.4.3</version>
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml:76]: Any reason to change this? Can you please add what errors you were getting?

File: cdap-data-fabric/pom.xml [hunk 2]
```diff
@@ -45,13 +45,6 @@
     </dependency>
     <dependency>
       <groupId>co.cask.cdap</groupId>
-      <artifactId>cdap-common</artifactId>
-      <version>${project.version}</version>
-      <type>test-jar</type>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>co.cask.cdap</groupId>
```
Comment (sagarkapare) [inline cdap-data-fabric/pom.xml:54]: similarly can you please add why is this change required?

File: cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml (hunk 1 above)
Comment (awholegunch) [reply cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml:76]: If I leave this change in, `maven` complains:
```
[WARNING] Some problems were encountered while building the effective model for co.cask.cdap:cdap-etl-tools:jar:3.5.5
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-shade-plugin is missing. @ co.cask.cdap:cdap-etl-tools:[unknown-version], /Users/john/Source/cdap_3.5/cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml, line 73, column 15
```
and then the build eventually hangs.

File: cdap-data-fabric/pom.xml (hunk 2 above)
Comment (awholegunch) [reply cdap-data-fabric/pom.xml:54]: This is duplication of lines 39 to 45. If I leave this change in, `maven` complains:

```
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for co.cask.cdap:cdap-data-fabric:jar:3.5.5
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: co.cask.cdap:cdap-common:test-jar -> duplicate declaration of version ${project.version} @ co.cask.cdap:cdap-data-fabric:[unknown-version], /Users/john/Source/cdap_3.5/cdap-data-fabric/pom.xml, line 46, column 17
[WARNING] 

```
and then the build eventually hangs.

File: cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml (hunk 1 above)
Comment (awholegunch) [reply cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml:76]: I copied the version from another pom.xml. It's the highest. Open to other suggestions.

Comment (sreevatsanraman) [reply cdap-app-templates/cdap-etl/cdap-etl-tools/pom.xml:76]: Let's open a JIRA to update shade plugin version to 2.4.3 everywhere.

**Comment by sreevatsanraman** [comment]
Open a JIRA to use consistent version of shade plugin and merge this PR. LGTM

**Comment by awholegunch** [comment]
Created https://issues.cask.co/browse/CDAP-9332

File: cdap-docs/build.sh [hunk 3]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
     local debug_flag=''
   fi
   if [[ ${javadoc_type} == ${ALL} ]]; then
-    javadoc_run="javadoc:aggregate"
+    javadoc_run="javadoc:aggregate -P release"
   else
-    javadoc_run="site"
+    javadoc_run="site -P templates"
```
Comment (wolf31o2) [inline cdap-docs/build.sh:193]: Why are we changing how this worked? We're mixing commands and profiles. If this is necessary, split these into separate variables... better to leave them as they were with the other changes in these comments.

File: cdap-docs/build.sh [hunk 4]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
...
   local start=`date`
   cd ${PROJECT_PATH}
   MAVEN_OPTS="-Xmx4g -XX:MaxPermSize=256m" # match other CDAP builds
-  mvn clean package ${javadoc_run} -P examples,templates,release -DskipTests -Dgpg.skip=true -DisOffline=false ${debug_flag}
-  warnings=$?
-  if [[ ${warnings} -eq 0 ]]; then
-    echo
-    echo "Javadocs Build Start: ${start}"
-    echo "                 End: `date`"
+  local temp_repo="-Dmaven.repo.local=/tmp/repo-cdap-docs/${GIT_BRANCH}"
```
Comment (wolf31o2) [inline cdap-docs/build.sh:198]: Do not use something outside the build location. Put this somewhere within the build directory, not `/tmp`... somewhere under `target` so it gets cleaned on a clean.

File: cdap-docs/build.sh [hunk 5]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
...
+  echo "Using temp_repo '${temp_repo}'"
+  if [[ -d ${temp_repo} ]]; then
+    rm -rf ${temp_repo}
+  fi
```
Comment (wolf31o2) [inline cdap-docs/build.sh:202]: If you stick this under target somewhere, it shouldn't ever exist on a clean checkout, anyway.

File: cdap-docs/build.sh [hunk 6]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
...
+  local temp_repo="-Dmaven.repo.local=/tmp/repo-cdap-docs/${GIT_BRANCH}"
+  echo "Using temp_repo '${temp_repo}'"
+  if [[ -d ${temp_repo} ]]; then
+    rm -rf ${temp_repo}
+  fi 
+  echo "========================================================"
+  echo "Building and installing CDAP to ${temp_repo}"
+  echo "command:"
+  echo "mvn clean install ${temp_repo} -P examples,templates,release -DskipTests -Dgpg.skip=true"
+  mvn clean install ${temp_repo} -P examples,templates,release -DskipTests -Dgpg.skip=true
```
Comment (wolf31o2) [inline cdap-docs/build.sh:207]: So you're duplicating an entire CDAP build, including downloading *ALL* dependencies, with no cache, on every build? This is excessive and increases bandwidth usage dramatically. Our dependency JARs should already be cached in our local Maven repository. We should use that, instead.

```
mkdir -p ${temp_repo}
if [[ -d ${HOME}/.m2/repository ]]; then
  cp -a ${HOME}/.m2/repository/* ${temp_repo}
  # Cleanup any CDAP JARs, so we use our own
  rm -rf ${temp_repo}/co/cask/cdap
fi
```

File: cdap-docs/build.sh [hunk 7]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
...
+  errors=$?
+  if [[ ${errors} -eq 0 ]]; then
+    echo "========================================================"
+    echo "Building and installing CDAP finished; running Javadocs"
+    echo "command:"
+    echo "mvn ${javadoc_run} ${temp_repo} -DskipTests -Dgpg.skip=true -DisOffline=false ${debug_flag}"
+    mvn ${javadoc_run} ${temp_repo} -DskipTests -Dgpg.skip=true -DisOffline=false ${debug_flag}
```
Comment (wolf31o2) [inline cdap-docs/build.sh:214]: This Maven command should be rolled into the above Maven command.

File: cdap-docs/build.sh [hunk 8]
```diff
@@ -182,24 +188,43 @@ function build_javadocs() {
...
+    errors=$?
```
Comment (wolf31o2) [inline cdap-docs/build.sh:215]: This shadows the outer errors. Don't use the same variable name for multiple instances.

Comment (wolf31o2) [reply cdap-docs/build.sh:215]: Also, this won't be necessary if the commands are rolled together.

**Comment by wolf31o2** [comment]
Ugh... commented on the wrong PR.

File: cdap-docs/build.sh (hunk 5 above)
Comment (awholegunch) [reply cdap-docs/build.sh:202]: Removed.

File: cdap-docs/build.sh (hunk 6 above)
Comment (awholegunch) [reply cdap-docs/build.sh:207]: Added.

File: cdap-docs/build.sh (hunk 4 above)
Comment (awholegunch) [reply cdap-docs/build.sh:198]: Moved to ${TARGET_PATH}

File: cdap-docs/build.sh (hunk 3 above)
Comment (awholegunch) [reply cdap-docs/build.sh:193]: Fixed.

File: cdap-docs/build.sh (hunk 7 above)
Comment (awholegunch) [reply cdap-docs/build.sh:214]: Fixed.

File: cdap-docs/build.sh (hunk 8 above)
Comment (awholegunch) [reply cdap-docs/build.sh:215]: Fixed.

**Comment by awholegunch** [comment]
I will start a new branch, PR, and reference these comments from it. Once that is working and approved, use it to update https://github.com/caskdata/cdap/pull/8554 and https://github.com/caskdata/cdap/pull/8555.
